File: app/src/main/java/com/agrogames/islandsofwar/factories/GraphicsFactory.java
package com.agrogames.islandsofwar.factories;

import android.util.Log;

import com.agrogames.islandsofwar.engine.abs.another.IGraphicsObject;
import com.agrogames.islandsofwar.engine.abs.common.Point;
import com.agrogames.islandsofwar.engine.impl.another.GraphicsObject;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

public class GraphicsFactory {

    public static final List<JSONObject> graphics = new ArrayList<>();
    public static IGraphicsObject getGraphics(String name, Point location, float rotation){
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
            try {
                return graphics(getJSONGraphics(name), location, rotation);
            } catch (JSONException e) {
                e.printStackTrace();
                Log.e("IOW","Error in JSON schema", e);
            }
        }
        return new GraphicsObject("ERROR", new Point(-1, -1), 0, 0);
    }

    private static JSONObject getJSONGraphics(String name){
        try {
            for (JSONObject g : graphics){
                if(g.getString("name").equals(name)){
                    return g;
                }
            }
            Log.e("IOW","Graphics object with name " + name + " NOT FOUND");
            return new JSONObject();
        } catch (JSONException e) {
            e.printStackTrace();
            Log.e("IOW","Error in JSON syntax", e);
            return new JSONObject();
        }
    }

    private static IGraphicsObject graphics(JSONObject json, Point location, float rotation) throws JSONException{
        if(json.has("next")){
            return new GraphicsObject("graphics/" + json.getString("name"), location,
                    rotation,
                    getGraphics(json.getString("next"), location, rotation),
                    (float) json.getDouble("lifetime"));
        }
        return new GraphicsObject("graphics/" + json.getString("name"), location,
                rotation,
                (float) json.getDouble("lifetime"));
    }
}

File: app/src/main/java/com/agrogames/islandsofwar/factories/GraphicsFactory.kt
package com.agrogames.islandsofwar.factories

import com.agrogames.islandsofwar.engine.abs.common.Point
import com.agrogames.islandsofwar.engine.abs.graphics.IGraphicsObject
import com.agrogames.islandsofwar.engine.impl.graphics.GraphicsObject
import org.json.JSONObject

object GraphicsFactory {
    val graphics: MutableList<JSONObject> = ArrayList()
    fun getGraphics(name: String, location: Point, rotation: Float): IGraphicsObject =
        graphics(getJSONGraphics(name), location, rotation)

    private fun getJSONGraphics(name: String): JSONObject =
        graphics.find { it.getString("name") == name }!!

    private fun graphics(json: JSONObject, location: Point, rotation: Float): IGraphicsObject =
        if (json.has("next")) {
            GraphicsObject(
                    "graphics/" + json.getString("name"),
                    location,
                    rotation,
                    json.optString("spawn_sound"),
                    json.getDouble("lifetime").toFloat(),
                    getGraphics(json.getString("next"), location, rotation))
        } else GraphicsObject("graphics/" + json.getString("name"),
                location,
                rotation,
                json.optString("spawn_sound"),
            json.getDouble("lifetime").toFloat())
}